### ISSUE TITLE
Support Addon Coverage

### ISSUE THREAD

**Issue by offirgolan** [kategengler/ember-cli-code-coverage#14]
As of currently (v0.2.0) addon support for test coverage is not present. After testing it with my addon, Im only getting coverage for `app/resolver.js` which is pretty meaningless. Im opening up this issue to track the progress for this. Once this lands, ill go ahead and replace blanketjs with this in ember-addon-genie.

**Comment by kategengler** [comment]
This is my top priority for this addon, I have a WIP but was waiting for that PR that landed earlier today. That said, I will likely not get to it until July 4 weekend at the earliest.

**Comment by rwwagner90** [comment]
@kategengler do you have a branch we could look at? If I have some free time, I may be able to help out.

**Comment by rwwagner90** [comment]
Just let me know if you want me to take a stab at any of this. So pumped for this addon 😄

**Comment by rwjblue** [comment]
@rwwagner90 - I did some spike work the other day, but ran out of time to finish the other night (honestly I fell asleep 💤 ). Maybe you could decipher and try to bring it home?  https://github.com/kategengler/ember-cli-code-coverage/compare/addon-spike

**Comment by bmeurant** [comment]
I tested it on one of my addon and it seemed working perfectly :+1: 
I made [some simple changes](https://github.com/bmeurant/ember-cli-code-coverage/commit/2d50b3c0426ae4c795af01742af265fb959042ff) on my fork regarding tests.

Merging to home should be simple but I am not sure to see all cases and impacts. It seems pretty good to me but I am not mastering addons & hooks and I am sure I miss something ...

@rwjblue I am curious, what is the state of this for you ?

**Comment by elwayman02** [comment]
@kategengler does the addon spike play nicely with ember-try?

**Comment by bmeurant** [comment]
@elwayman02 @kategengler I tested it running a `ember try:each` and I did not experiment any specific issue

**Comment by bmeurant** [comment]
I found some incompatibilities between addon-spike implementation and usage of eslint in the covered addon.

When running tests with coverage enabled, ember-cli-eslint fails as described in ember-cli/ember-cli-eslint#56. 

I suspect the call to `preprocessTree` in `included` [here](https://github.com/bmeurant/ember-cli-code-coverage/blob/addon-spike/index.js#L28). This seams to lead to multiple mergeTrees because of the call to `BroccoliMergeTree` [here](https://github.com/bmeurant/ember-cli-code-coverage/blob/addon-spike/index.js#L79)

You can find an example of this on this [dedicated branch of my addon](https://github.com/bmeurant/ember-array-contains-helper/tree/cli-code-coverage-addon)

Sadly, I have no idea how to fix that or even where to start. I suppose that we have to find another way to include addon files in `preprocessTree` or something similar. But I fear that my knowledge of broccoli and ember-cli is still too light for now :confused:

**Comment by arjansingh** [comment]
:+1: Love this addon. Can someone give me a lay of the land for what needs to be done to make this work with addons? (@rwjblue?) 

If I can get a high level overview of what needs needs changing or how an addon's structure is different from an app's, I'm happy to take a stab at addon compatibility.

**Comment by rwjblue** [comment]
I believe that @bmeurant has been working in cleaning up the spike I linked above.

**Comment by bmeurant** [comment]
@rwjblue yes I did some (very light) cleanup on your initial work ([here](https://github.com/bmeurant/ember-cli-code-coverage/tree/addon-spike)). The spike basically works (congrats!). But, as said before, I was stuck because of some incompatibilities with ember-cli-eslint (``included`` call to ``preprocessTree`` is leading to multiple merge trees and this is not supported by ember-cli-eslint.

I have still no good knowledge of broccoli & ember-cli internals and I did not have enough time to go deeper to find an alternative. I take any advice. Or @arjansingh, if you want to investigate further, feel free!

BTW, I see now that I generated a lot of noise here with some rebase/squash on local branches, sorry for that :confused:
